src/compile.c

```diff
@@ -702,13 +726,139 @@ static block bind_matcher(block matcher, block body) {
   return BLOCK(matcher, body);
 }
 
+/* Build a wrapper around a destructuring matcher so that we can chain them
+ * when we have errors.  The approach is as follows:
+ * FORK_OPT NEXT_MATCHER (unless last matcher)
+ * existing_matcher_block
+ * restructure matched variables into a new jv_object
+ * JUMP END
+ *
+ * END will destructure the object returned from a wrapper into all of the
+ * variables contained in the entire destructuring expression and is what the
+ * program's body binds to.
+ */
+static block alternation_wrapper(block submatcher, block *final_matcher, int is_first, int is_last, jv vars_list) {
+    // We need to generate an output object of what succeeded here so we can
+    // bind everyone to those values.
+    block end = gen_subexp(gen_const(jv_object()));
+
+    // Manually restructure the values we extracted into a new object.
+    jv_array_foreach(vars_list, _, var) {
+      end = BLOCK(end, gen_dictpair(gen_const(var), gen_op_unbound(LOADV, jv_string_value(var))));
+    }
+    block_append(&end, gen_op_simple(POP));
+    submatcher = bind_matcher(submatcher, end);
+
+    // If we're successful, jump to the end of the matchers
+    if (final_matcher) {
+      submatcher = BLOCK(submatcher, gen_op_target(JUMP, *final_matcher));
+    }
+
+    // Get rid of the error if we aren't the first matcher
+    if (!is_first) {
+      submatcher = BLOCK(gen_op_simple(POP), gen_op_simple(DUP), submatcher);
+    }
+
+    // If we aren't building the last submatcher, FORK_OPT to the end
+    // of this submatcher so we can skip to the next one on error
+    if (!is_last) {
+      submatcher = BLOCK(gen_op_target(FORK_OPT, submatcher), submatcher);
+    }
+    return submatcher;
+}
+
+// Extract destructuring var names from the block
+// *all_vars should be a jv_object (for set semantics)
+// *matcher_vars should be a jv_array
+static void alternation_vars(block b, jv *all_vars, jv *matcher_vars) {
```

Comment:
Though if you can make the change above re `STOREV`s, then you don't need the `matcher_vars` output here, just the `all_vars` output, which then can be an input and then this function can just return the modified object.  Then this function would indeed be generic.